### PR TITLE
Introduce optional company name and sales tax id columns to customer addresses

### PR DESCRIPTION
**Pull request by thyseus**
Feature:

For B2B (Business to Business) shops it is common to have the company_name and vat_id provided by the customers. With this Pull request the customers are able to provide it, if they want. It is optional.

I also added a little bit to the testing infrastructure:

Added a script bin/test.sh that migrates the test database and executes all necessary tests. Ensure to have the database bagisto_test at your mysql server as provided in the .env.testing file.

And some factories for easier testing.

### CODE REVIEW

File: packages/Webkul/Customer/src/Models/CustomerAddress.php [hunk 1]
```diff
@@ -8,5 +9,18 @@ class CustomerAddress extends Model implements CustomerAddressContract
...
     protected $table = 'customer_addresses';
 
-    protected $fillable = ['customer_id' ,'address1', 'country', 'state', 'city', 'postcode', 'phone', 'default_address'];
+    protected $fillable = [
+        'customer_id',
+        'company_name',
+        'vat_id',
+        'address1',
+        'address2',
+        'address3',
```
Comment (SteffenMahler) [inline packages/Webkul/Customer/src/Models/CustomerAddress.php]: Hi Herbert,
there is no field "address3" in DB schema. So remove this from fillable or add a migration to alter the table.

Comment (thyseus) [reply packages/Webkul/Customer/src/Models/CustomerAddress.php]: Correct. I have removed the line.

Comment (SteffenMahler) [reply packages/Webkul/Customer/src/Models/CustomerAddress.php]: Nice.
I would also suggest to split the existing field "name" in table customer_addresses into "first_name" and "last_name" as it is in table customers.

Happy coding.

File: packages/Webkul/Shop/src/Resources/views/customers/account/address/edit.blade.php [hunk 2]
```diff
@@ -73,7 +94,7 @@
                     </div>
                 </div>
 
-            </form>
+            </formi>
```
Comment (SteffenMahler) [inline packages/Webkul/Shop/src/Resources/views/customers/account/address/edit.blade.php]: Typo

File: tests/functional/Customer/CustomerCest.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        $I->wantTo('Ensure that the company_name field is being displayed');
+        $I->seeElement('.account-table-content > div:nth-child(2) > input:nth-child(2)');
+
+        // we need to use this css selector to hit the correct <form>. There is another one at the
+        // page header (search)
+        $I->submitForm($formCssSelector, $this->fields);
+        $I->seeInSource('The given vat id has a wrong format');
+
+        // valid vat id:
+        $this->fields['vat_id'] = 'DE123456789';
```
Comment (SteffenMahler) [inline tests/functional/Customer/CustomerCest.php]: $faker->vat(false) to generate a valid vat without whitespace

File: packages/Webkul/Shop/src/Resources/views/customers/account/address/edit.blade.php (hunk 2 above)
Comment (thyseus) [reply packages/Webkul/Shop/src/Resources/views/customers/account/address/edit.blade.php]: Whooops. Corrected.

File: packages/Webkul/Customer/src/Repositories/CustomerAddressRepository.php [hunk 4]
```diff
@@ -63,16 +62,18 @@ public function update(array $data, $id)
 
         Event::fire('customer.addresses.update.before', $id);
 
-        if (isset($data['default_address']) ) {
+        if (isset($data['default_address'])) {
```
Comment (SteffenMahler) [inline packages/Webkul/Customer/src/Repositories/CustomerAddressRepository.php]: Please use your nice refactoring from line 36.

**Comment by thyseus** [comment]
@rahulshukla-webkul could you please provide a example in your codebase where a error message coming from the server-side is being displayed? 

Client-side validation is not enough in this case because the server is validating the vat_id.

I have found no example at http://vee-validate.logaretm.com/v2/guide/ and no example in bagisto so far.

**Comment by thyseus** [comment]
@rahulshukla-webkul Other approach: what do you think about removing the server side vat_id validation for now to get this PR merged?

**Comment by rahulshukla-webkul** [comment]
Hi @thyseus 

Remove this one - 

` @if ($errors->any())
                <div class="alert alert-danger">
                    <ul>
                        @foreach ($errors->all() as $error)
                            <li>{{ $error }}</li>
                        @endforeach
                    </ul>
                </div>
            @endif`

and create vat_id field like -  

`<div class="control-group {!! $errors->has('vat_id') ? 'has-error' : '' !!}">
                        <label for="vat_id">{{ __('shop::app.customer.account.address.create.vat_id') }}</label>
                        <input type="text" class="control" name="vat_id"  data-vv-as="&quot;{{ __('shop::app.customer.account.address.create.vat_id') }}&quot;">

                        <span class="control-error" v-if="{!! $errors->has('vat_id') !!}">
                            @foreach ($errors->get('vat_id') as $message)
                               {{ $message }}
                            @endforeach
                        </span>
                    </div>`

**Comment by thyseus** [comment]
@rahulshukla-webkul thanks a lot, will implement today !

**Comment by thyseus** [comment]
@rahulshukla-webkul ok, i implemented it this way: 

https://github.com/bagisto/bagisto/pull/2027/commits/7c9908401f73525c9729d633c2b44a15780025f5

please check if this is the intended way to do it :)

**Comment by bosix** [comment]
@thyseus please have a look at the merge conflicts. There are several in the OrderRepository

**Comment by thyseus** [comment]
> @thyseus please have a look at the merge conflicts. There are several in the OrderRepository

thanks, @bosix - conflicts resolved.

**Comment by bosix** [comment]
@thyseus your welcome. Please also have a look at the changes requested by @rahulshukla-webkul

**Comment by thyseus** [comment]
@rahulshukla-webkul - i merged master and updated my package.json

btw: i see you commitet some node_modules folder into master. i highely doubt that this is a wanted behavior. a rebase without the node_modules folders would save hundreds of megabytes of unnecessary code in the master (as soon as possible)

**Pull request merged**